The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11177722316/job/31073799345) [(plain)](https://github.com/rust-lang/rust/commit/dd7dfe8d63b76730085dc7f2be6dd9be2d7b89dc/checks/31073799345/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.781 Building wheels for collected packages: reuse
#13 2.782   Building wheel for reuse (pyproject.toml): started
#13 3.030   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.031   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#13 3.031   Stored in directory: /tmp/pip-ephem-wheel-cache-z7q8qir0/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.034 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.427 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.428 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 3.982 Collecting virtualenv
#13 3.982 Collecting virtualenv
#13 4.038   Downloading virtualenv-20.26.6-py3-none-any.whl (6.0 MB)
#13 4.311      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.0/6.0 MB 22.1 MB/s eta 0:00:00
#13 4.373 Collecting filelock<4,>=3.12.2
#13 4.382   Downloading filelock-3.16.1-py3-none-any.whl (16 kB)
#13 4.403 Collecting distlib<1,>=0.3.7
#13 4.412   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 4.426      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 39.9 MB/s eta 0:00:00
#13 4.463 Collecting platformdirs<5,>=3.9.1
#13 4.472   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#13 4.553 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.759 Successfully installed distlib-0.3.8 filelock-3.16.1 platformdirs-4.3.6 virtualenv-20.26.6
#13 DONE 4.8s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      219072 kB
DirectMap2M:     9218048 kB
DirectMap1G:     9437184 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://static.rust-lang.org/dist/2024-09-22/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/2024-09-22/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt
---
    Finished `release` profile [optimized] target(s) in 31.61s
##[endgroup]
fmt check
fmt: checked 5604 files
Diff in /checkout/src/bootstrap/src/core/builder.rs:1696:
                         .arg("-")
                         .run_capture(self)
                         .stderr();
-                    let not_supported = error.lines().any(|line| {
-                        line.contains("unsupported crate type `proc-macro`")
+                    let not_supported = error
+                        .lines()
+                        .lines()
+                        .any(|line| line.contains("unsupported crate type `proc-macro`"));
                     if !not_supported {
                         cargo.arg("-Zdual-proc-macros");
                         rustflags.arg("-Zdual-proc-macros");
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/src/tools/unicode-table-generator/src/range_search.rs" "/checkout/src/tools/unicode-table-generator/src/case_mapping.rs" "/checkout/src/tools/unicode-table-generator/src/main.rs" "/checkout/src/tools/unicode-table-generator/src/cascading_map.rs" "/checkout/src/tools/unicode-table-generator/src/raw_emitter.rs" "/checkout/src/tools/unicode-table-generator/src/skiplist.rs" "/checkout/src/tools/unicode-table-generator/src/unicode_download.rs" "/checkout/src/tools/jsondocck/src/config.rs" "/checkout/src/tools/jsondocck/src/main.rs" "/checkout/src/tools/jsondocck/src/error.rs" "/checkout/src/tools/jsondocck/src/cache.rs" "/checkout/src/tools/compiletest/src/common.rs" "/checkout/src/tools/compiletest/src/main.rs" "/checkout/src/tools/compiletest/src/runtest.rs" "/checkout/src/tools/compiletest/src/util.rs" "/checkout/src/tools/compiletest/src/errors/tests.rs" "/checkout/src/tools/compiletest/src/header/test-auxillary/known_directive.rs" "/checkout/src/tools/compiletest/src/header/test-auxillary/unknown_directive.rs" "/checkout/src/tools/compiletest/src/header/test-auxillary/error_annotation.rs" "/checkout/src/tools/compiletest/src/header/test-auxillary/known_legacy_directive.rs" "/checkout/src/tools/compiletest/src/header/needs.rs" "/checkout/src/tools/compiletest/src/header/tests.rs" "/checkout/src/tools/compiletest/src/header/cfg.rs" "/checkout/src/tools/compiletest/src/tests.rs" "/checkout/src/tools/compiletest/src/errors.rs" "/checkout/src/tools/compiletest/src/read2.rs" "/checkout/src/tools/compiletest/src/runtest/rustdoc.rs" "/checkout/src/tools/compiletest/src/runtest/js_doc.rs" "/checkout/src/tools/compiletest/src/runtest/ui.rs" "/checkout/src/tools/compiletest/src/runtest/codegen.rs" "/checkout/src/tools/compiletest/src/runtest/coverage.rs" "/checkout/src/tools/compiletest/src/runtest/run_make.rs" "/checkout/src/tools/compiletest/src/runtest/debugger.rs" "/checkout/src/tools/compiletest/src/runtest/crashes.rs" "/checkout/src/tools/compiletest/src/runtest/tests.rs" "/checkout/src/tools/compiletest/src/runtest/valgrind.rs" "/checkout/src/tools/compiletest/src/runtest/incremental.rs" "/checkout/src/tools/compiletest/src/runtest/debuginfo.rs" "/checkout/src/tools/compiletest/src/runtest/assembly.rs" "/checkout/src/tools/compiletest/src/runtest/rustdoc_json.rs" "/checkout/src/tools/compiletest/src/runtest/codegen_units.rs" "/checkout/src/tools/compiletest/src/runtest/mir_opt.rs" "/checkout/src/tools/compiletest/src/runtest/pretty.rs" "/checkout/src/tools/compiletest/src/util/tests.rs" "/checkout/src/tools/compiletest/src/raise_fd_limit.rs" "/checkout/src/tools/compiletest/src/compute_diff.rs" "/checkout/src/tools/compiletest/src/json.rs" "/checkout/src/tools/compiletest/src/header.rs" "/checkout/src/tools/compiletest/src/command-list.rs" "/checkout/src/tools/compiletest/src/lib.rs" "/checkout/src/tools/compiletest/src/read2/tests.rs" "/checkout/src/bootstrap/src/core/mod.rs" "/checkout/src/bootstrap/src/core/metadata.rs" "/checkout/src/bootstrap/src/core/download.rs" "/checkout/src/bootstrap/src/core/builder.rs" "/checkout/src/bootstrap/src/core/builder/tests.rs" "/checkout/src/bootstrap/src/core/config/mod.rs" "/checkout/src/bootstrap/src/core/config/config.rs" "/checkout/src/bootstrap/src/core/config/flags.rs" "/checkout/src/bootstrap/src/core/config/tests.rs" "/checkout/src/bootstrap/src/core/build_steps/check.rs" "/checkout/src/bootstrap/src/core/build_steps/suggest.rs" "/checkout/src/bootstrap/src/core/build_steps/mod.rs" "/checkout/src/tools/miropt-test-tools/src/lib.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Fri Oct  4 09:50:26 UTC 2024
  network time: Fri, 04 Oct 2024 09:50:26 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

--- a/compiler/rustc_driver_impl/src/lib.rs
+++ b/compiler/rustc_driver_impl/src/lib.rs
@@ -29,13 +29,12 @@ use std::path::PathBuf;
 use std::process::{self, Command, Stdio};
 use std::sync::atomic::{AtomicBool, Ordering};
 use std::sync::{Arc, OnceLock};
-use std::time::{Duration, Instant, SystemTime};
+use std::time::{Instant, SystemTime};
 use std::{env, str};
 
 use rustc_ast as ast;
 use rustc_codegen_ssa::traits::CodegenBackend;
 use rustc_codegen_ssa::{CodegenErrors, CodegenResults};
-use rustc_const_eval::CTRL_C_RECEIVED;
 use rustc_data_structures::profiling::{
     TimePassesFormat, get_resident_set_size, print_time_passes_entry,
 };
@@ -1577,8 +1576,8 @@ pub fn install_ctrlc_handler() {
         // time to check CTRL_C_RECEIVED and run its own shutdown logic, but after a short amount
         // of time exit the process. This sleep+exit ensures that even if nobody is checking
         // CTRL_C_RECEIVED, the compiler exits reasonably promptly.
-        CTRL_C_RECEIVED.store(true, Ordering::Relaxed);
-        std::thread::sleep(Duration::from_millis(100));
+        rustc_const_eval::CTRL_C_RECEIVED.store(true, Ordering::Relaxed);
+        std::thread::sleep(std::time::Duration::from_millis(100));
         std::process::exit(1);
     })
     .expect("Unable to install ctrlc handler");

--- a/src/bootstrap/src/core/builder.rs
+++ b/src/bootstrap/src/core/builder.rs
@@ -1683,10 +1683,24 @@ impl<'a> Builder<'a> {
         match mode {
             Mode::Std | Mode::ToolBootstrap | Mode::ToolStd => {}
             Mode::Rustc | Mode::Codegen | Mode::ToolRustc => {
-                // Build proc macros both for the host and the target
+                // Build proc macros both for the host and the target unless proc-macros are not
+                // supported by the target.
                 if target != compiler.host && cmd_kind != Kind::Check {
-                    cargo.arg("-Zdual-proc-macros");
-                    rustflags.arg("-Zdual-proc-macros");
+                    let error = command(self.rustc(compiler))
+                        .arg("--target")
+                        .arg(target.rustc_target_arg())
+                        .arg("--print=file-names")
+                        .arg("--crate-type=proc-macro")
+                        .arg("-")
+                        .run_capture(self)
+                        .stderr();
+                    let not_supported = error
+                        .lines()
+                        .any(|line| line.contains("unsupported crate type `proc-macro`"));
+                    if !not_supported {
+                        cargo.arg("-Zdual-proc-macros");
+                        rustflags.arg("-Zdual-proc-macros");
+                    }
                 }
             }
         }

